### ISSUE TITLE
Deadlocking issue re-emerged

### ISSUE THREAD

**Issue by juslesan** [paullouisageneau/libdatachannel#619]
Hi @paullouisageneau and @murat-dogan !

Recently, I've run in to two separate cases in two different versions that appear to be deadlocks originating from libdatachannel in Streamr's network nodes. In the first case, an older version of node-datachannel `v0.2.3` is used and the process experiences rare deadlocks similiar to what was experienced around a year ago. As the version is older, we attempted to update node-datachannel to the latest version `v0.3.1`. However, the deadlocking issue appears to be more severe in the latest version causing the network nodes to no longer pass local integration tests without locking. Before stalling, this error messages is consistently being logged:

```
Test suite failed to run

    libdatachannel error# cleanup timeout (possible deadlock)

      73 |     },
      74 |     cleanUp(): void {
    > 75 |         nodeDataChannel.cleanup()
         |                         ^
      76 |     }
      77 | })
```

[This pull request](https://github.com/streamr-dev/network-monorepo/pull/605) in the Streamr Network's monorepo showcases the issue. [Here's an example test run.](https://github.com/streamr-dev/network-monorepo/runs/6274107520?check_suite_focus=true)

I did give a try at adding error handling around the clean up and removing it completely, but neither worked.

It would great to if you can take a look at this issue ASAP. Some network nodes that perform heavier tasks are occasionally deadlocking in production with the older versions and the library cannot be updated to latest if integration tests are deadlocking.

If you need more info or help I'm available.

Thanks!

**Comment by murat-dogan** [comment]
Hi,
I can re-produce the issue. 
You are calling `nodeDataChannel.cleanup`  multiple times by calling `NodeToNode.stop`. Is that normal?

Changing the test like this solves the issue.
```
afterAll(async () => {
        await Promise.allSettled([
            tracker.stop(),
            contactNode.stop(),
            otherNodes.map((node) => node.stop()),
            cleanup()
        ])
    })
```

Of course we should remove cleanup code from `NodeToNode.stop`.

Is that a solution for you?

**Comment by murat-dogan** [comment]
Wait, that's not true. It is not making a global cleanup.

**Comment by teogeb** [comment]
Hi! Thanks for the help. We have made some progress :)

We implemented reference counting to ensure `nodeDataChannel.cleanup()` get called only once. That seems to solve most of the deadlock issues. https://github.com/streamr-dev/network-monorepo/pull/605/commits/b414536354065fefe3ca30364ae5daf3a70ff28f

However we have another issue that comes up with running  `npx jest test/integration/webrtc-error-scenarios-during-signalling.test.ts` or `npx jest test/integration/proxy-connections.test.ts`

It will either freezes or segfaults with one of the errors:

```
node: tpp.c:83: __pthread_tpp_change_priority: Assertion `new_prio == -1 || (new_prio >= fifo_min_prio && new_prio <= fifo_max_prio)' failed.
[1]    4690 IOT instruction (core dumped)  npx jest test/integration/proxy-connections.test.ts
```

```
node: pthread_mutex_lock.c:94: ___pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

```
terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument
```

We can try to write a more isolated repro if it helps.

**Comment by murat-dogan** [comment]
Good to hear that.
Please note that I can not see "Global cleanup" message from `libdatachannel`. Is it a problem @paullouisageneau ?

The new error is related to this PR https://github.com/murat-dogan/node-datachannel/pull/105#pullrequestreview-963254500.@

I can re-produce the error with 0.3.1 but it seems solved with the PR.
As @paullouisageneau  said he is working on a better re-factoring, I am waiting for it or maybe I can improve it.

https://github.com/murat-dogan/node-datachannel/issues/103#issuecomment-1118474145

**Comment by paullouisageneau** [comment]
Thanks for reporting, I'm having a look.
 
> Please note that I can not see "Global cleanup" message from libdatachannel. Is it a problem @paullouisageneau ?

It could be if the rest of the debug log is displayed.

**Comment by paullouisageneau** [comment]
> I guess these segfaults may already be fixed by the bullet points you outlined? However just wanted to clarify that these occur even without calls to `nodeDataChannel.cleanup()`.

That's right, the segfaults were definitely caused by the first issue, which is unrelated to `nodeDataChannel.cleanup()`.

The second issue triggered by `nodeDataChannel.cleanup()` is fixed by https://github.com/paullouisageneau/libdatachannel/pull/621.

**Comment by murat-dogan** [comment]
I have release v0.3.2 with latest fixes.
Could you please try with the new version?
